Add ConversionTaskScheduler

marl0vv opened the pull request:
Добавил класс, который по расписанию:
- достаёт из БД все PENDING записи
- конвертирует изображения
- сохраняет конвертированные изображения в out folder
- удаляет исходные изображения
- пишет в БД об успешной конвертации и сохранённом файле

marl0vv committed:
Add task creation date and converted name to table

marl0vv committed:
Add Scheduler

Add scheduler that converts PENDING images once in 10 seconds

RavenVSS commented on src/main/java/com/glamik/webpconverter/service/ConversionTaskService.java:
```diff
@@ -6,21 +6,41 @@
                 .save(conversionTask);
     }
 
+    public void processConversionTask(ConversionTask task) throws IOException {
```
Вынести в отдельный класс, например в Command

RavenVSS commented on src/main/java/com/glamik/webpconverter/scheduler/ConversionTaskScheduler.java:
```diff
@@ -0,0 +1,29 @@
+    public void processPendingConversionTasks() throws IOException {
+        List<ConversionTask> pendingTasks = conversionTaskRepository.findByStatus(ConversionTaskStatus.PENDING);
+        for (ConversionTask task : pendingTasks) {
+            conversionTaskService.processConversionTask(task);
```
Сделать проверку try catch, в случае ошибок менять статус на ERROR, в лог выводить информацию об ошибке и стектрейс (в lombok аннотация `@Slf4j`)  

Опционально: можно сохранить информацию об ошибке в БД

marl0vv committed:
Take out processing conversion task into separate Command class

Take out processing conversion task into separate Command class in order to keep up with Single-responsibility rule

marl0vv committed:
Add logs and error message in database

RavenVSS commented on src/main/java/com/glamik/webpconverter/command/ProcessConversionTaskCommand.java:
```diff
@@ -0,0 +1,50 @@
+            outputFile = fileService.saveOutputFile(outputFile);
+            Files.delete(filesystemFile.toPath());
+
+            task.setStatus(ConversionTaskStatus.SUCCESS);
```
Лучше вынести работу с репозиторием в сервис  

В сервисе сделать метод для изменения статуса, например:  
```
ConversionTaskService.setStatus(Long Id, ConversionTaskStatus status, ErrorMessage error);
```

RavenVSS commented on src/main/java/com/glamik/webpconverter/repository/ConversionTaskRepository.java:
```diff
@@ -1,7 +1,13 @@
+
 public interface ConversionTaskRepository extends JpaRepository<ConversionTask, Long> {
+
+    List<ConversionTask> findByStatus(ConversionTaskStatus status);
```
Почитать: https://docs.spring.io/spring-data/jpa/reference/jpa/query-methods.html

marl0vv committed:
Take out working with repository to Service

RavenVSS commented on src/main/java/com/glamik/webpconverter/service/ConversionTaskService.java:
```diff
@@ -17,10 +25,22 @@ public ConversionTask saveConversionTask(String originalFileName, String filesys
                 .save(conversionTask);
     }
 
+    public void setConversionSuccessStatus(ConversionTask task, ConversionTaskStatus status, String convertedName) {
```
Добавь `@Transactional` и запрос по Id

marl0vv committed:
Add @Transactional to methods working with repository
